src/Workspaces/Core/Portable/Workspace/Host/Status/ISolutionStatusService.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    /// <summary>
+    /// Provides solution status
+    /// </summary>
+    internal interface ISolutionStatusService : IWorkspaceService
+    {
+        /// <summary>
+        /// Wait until given solution and its children are fully loaded.
+        /// </summary>
+        Task WaitForAsync(Solution solution, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Wait until given project and its children are fully loaded.
+        /// </summary>
+        Task WaitForAsync(Project project, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Indicates whether given solution and its children are fully loaded or not. 
+        /// </summary>
+        Task<bool> IsFullyLoadedAsync(Solution solution, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Indicates whether given project and its children are fully loaded or not.
+        /// </summary>
+        Task<bool> IsFullyLoadedAsync(Project project, CancellationToken cancellationToken);
```

Comment:
where is 'Fully loaded' defined?

Reply:
I guess question for @cartermp

Reply:
"Fully loaded" means that a project has DTB info - resolved references, packages, etc. In other words, Roslyn isn't missing information it needs to be able to do stuff that people would expect, like offer completion on `Console.` when the `System` namespace is in scope.

Things are more complicated than that as far as the project system is concerned, but CPS (and eventually Legacy) should be able to hydrate an API that the VS Platform is providing so that we can have a binary yes/no on if we've got the info we need.

Reply:
@cartermp Thanks.  That's very helpful.  Given that, i definitely think the internal Roslyn API has issues.  It should not be based on the immutable Solution/Project/Doc concepts.  It could be based either on just Workspace, along with potential fine-grained questions like "Is this solution-id in this workspace fully-loaded" or "is this project-id in this workspace fully loaded".